addons/survey/models/survey_user.py

```diff
@@ -381,13 +385,34 @@ def _check_answer_type(self):
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
-            suggested_answer_id = vals.get('suggested_answer_id')
-            if suggested_answer_id:
-                vals.update({'answer_score': self.env['survey.question.answer'].browse(int(suggested_answer_id)).answer_score})
+            self._update_answer_is_correct(vals)
         return super(SurveyUserInputLine, self).create(vals_list)
 
     def write(self, vals):
-        suggested_answer_id = vals.get('suggested_answer_id')
-        if suggested_answer_id:
-            vals.update({'answer_score': self.env['survey.question.answer'].browse(int(suggested_answer_id)).answer_score})
+        self._update_answer_is_correct(vals)
         return super(SurveyUserInputLine, self).write(vals)
+
+    def _update_answer_is_correct(self, vals):
```

Comment:
The `_get_line_answer_values` is our centralized method that returns the create values for the answer.

I would create a `_get_answer_score_values` method that does everything you do here and returns the computed values. Then call it in `_get_line_answer_values` with something like `vals.update(self._get_line_answer_values(question, answer, answer_type))`

That will NOT ensure that those values will be correctly recomputed if the records are modified directly through a form view on the backend, so we usually don't do it (it's safer the way you did it, in create/write), but it's ok for this use case, since answers are input on the frontend only. And I feel like it will reduce the complexity a lot.

Can you give it a try so we see how it looks?

Maybe make a backup of this method somewhere just in case.

Reply:
Your implementation results in much cleaner code and works with my testing.